x/reward/keeper/time_management.go

```diff
@@ -0,0 +1,256 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/reward/types"
+)
+
+// UpdateUnexpiredRewardsProgram called from begin blocker, starts/ends or expires rewards programs.
+func (k Keeper) UpdateUnexpiredRewardsProgram(ctx sdk.Context) {
+	rewardPrograms, err := k.GetUnexpiredRewardPrograms(ctx)
```

Comment:
I'm a little worried that we have to iterate over all reward programs in every begin blocker. It's probably not going to be an issue until there are lots of rewards programs, but I don't know what "lots" counts as.

Maybe as a separate issue, we should create an index/queue similar to what was done with the `x/authz` module.

Reply:
i agree on the concern . @Taztingo  has some results on it https://docs.google.com/spreadsheets/d/1dsqdnI999a7N3tsDAF7yrjpTd1vFI3bzJ8BOIINx2n0/edit#gid=0 with 100 rewards program running concurrently.

Reply:
I'll create a separate issue, similar to authz